### DISCUSSION TITLE
How to work with tags?

### DISCUSSION THREAD

**Discussion by mliebelt** [mliebelt/pgn-parser#48] [General]
I am at the moment trying to implement all rules in the [specification](https://github.com/mliebelt/pgn-spec-commented/blob/main/pgn-specification.md). By doing that, I found some flaws in the grammar implemented, and fixed some of them. With tags, it is a little bit different, and therefore I wanted to start a discussion about it.

## Possible Requirements

What will the people want to do with the pgn-parser? That may depend ...

* Some may implement something, where it is  essential that only valid tags are used.
* Some may just use some tags (e.g. the  [seven rooster tag](https://github.com/mliebelt/pgn-spec-commented/blob/main/pgn-specification.md#811-seven-tag-roster)) and ignore the rest.
* Some may use games from sources like lichess.org or others, which are creative in their use of the tags.

What are the options here?

* Find a superset of everything that will work for everyone. :x: There is no way to have this.
* Find a common set per usage, and provide for each set a grammar that works. But how to allow the use of that grammar then?
* Be strict in what you accept as values for the known tags, but liberal for the rest. :question: Maybe that is a solution for some ...
* Add additional sets of tags (like the ones used by lichess: `variant`, `white|blackRatingDiff`, `white|blackFideId`).

What is your opinion about it? I would like to implement something that is

* conforming to the spec (currently it is too liberal)
* useful in all circumstances the same time.

**Comment by MichaelLeeHobbs** [comment]
I'm on the liberal side of tags. I work a lot with DICOM; tags, and HL7; segments. Both support customer tags/segments which is very useful. One issue here is tags/segment collision, more on the HL7 side. For example ZDS is a common custom segment in HL7. As it's custom what's inside is not standardized and that's where the collision come in. All that aside though it's not an issue for you as you are the parser. I'll add an example latter on how to post process that data and filter out non-standard tags.

**Reply by mliebelt** [reply]
I think the current situation is very liberal. It works like this:

1. The parser tries to read the tags including the expected values.
2. If that fails, it notice that failure as message, and reads the tag value as a string without evaluating it.
3. Tags that are not defined are read as is, and a message is emitted as well.

So if there is (later) the need to filter based on messages, or to report wrong tags, it is all in the data, you just have to read and report it.

**Reply by MichaelLeeHobbs** [reply]
Agree, it works well. I don't think you need to actually change anything was my point.
